# 异常

异常可以从请求处理程序中引发，并由Sanic自动处理。异常将消息作为第一个参数，也可以将状态代码作为HTTP响应传递回去。

## 使用

要抛出异常，只需从`sanic.exceptions`模块引发相关异常。

```python
from sanic.exceptions import SanicException

@app.route("/youshallnotpass")
async def no_no(request):
        raise SanicException("Something went wrong.", status_code=501)
```

常用异常

```python
# BadRequest 400
# Unauthorized 401
# Forbidden 403
# NotFound 404
# ServerError 500

from sanic import exceptions

@app.route("/login")
async def login(request):
    user = await some_login_func(request)
    if not user:
        raise exceptions.NotFound(f"Could not find user with username={request.json.username}")
```

## 特性

| name          | desc                |
| ------------- | ------------------- |
| `message`     | 信息                |
| `status_code` | 状态值              |
| `quiet`       | 是否写入`error_log` |
| `headers`     | 响应头              |
| `content`     | 内容                |
| `extra`       | 其他                |

## 处理

要覆盖Sanic对异常的默认处理，请使用`@app.exception`装饰器。装饰器期望将异常列表作为参数处理。您可以通过`SanicException`来捕获它们！装饰的异常处理程序函数必须将`Request`和`Exception`对象作为参数。

`app.error_handler`

```python
from sanic import Sanic

async def server_error_handler(request, exception):
    return text("Oops, server error", status=500)

app = Sanic()
app.error_handler.add(Exception, server_error_handler)
```

装饰器

```python
from sanic.exceptions import NotFound

@app.exception(NotFound, SomeCustomException)
async def ignore_404s(request, exception):
    return text("Yep, I totally found the page: {}".format(request.url))
```

## 内置处理

可以在每个路由中设置内置处理异常格式

```python
# HTML/JSON/text
@app.request("/", error_format="text")
async def handler(request):
    ...

```

## 自定义处理

在某些情况下，您可能希望在默认设置的基础上添加更多的错误处理功能。在这种情况下，您可以将Sanic的默认错误处理程序作为子类

```python
from sanic.handlers import ErrorHandler

class CustomErrorHandler(ErrorHandler):
    def default(self, request: Request, exception: Exception) -> HTTPResponse:
        ''' handles errors that have no error handlers assigned '''
        # You custom error handling logic...
        status_code = getattr(exception, "status_code", 500)
        return json({
          "error": str(exception),
          "foo": "bar"
        }, status=status_code)

app.error_handler = CustomErrorHandler()

```

## 反馈处理

```
app.config.FALLBACK_ERROR_FORMAT = "html"
app.config.FALLBACK_ERROR_FORMAT = "text"
app.config.FALLBACK_ERROR_FORMAT = "json"
app.config.FALLBACK_ERROR_FORMAT = "auto"
```

## 错误报告

Sanic有一个信号，允许您挂入异常报告过程。如果您想向Sentry或Rollbar等第三方服务发送异常信息，这很有用。这可以通过附加一个错误报告处理程序来方便地实现，如下所示：

```python
@app.report_exception
async def catch_any_exception(app: Sanic, exception: Exception):
	print("Caught exception:", exception)
```

